saga/README.md

```diff
@@ -0,0 +1,121 @@
+
+Similarly to the outbox pattern, this implementation avoids unsafe dual writes to a service's database and Apache Kafka by channeling all outgoing messages through the originating service's database and capturing them from there using CDC and Debezium.
+
+There are three services involved:
+
+* _order-service:_ originator and orchestrator of the Saga
+* _customer-service:_ approves or rejects the customer's credit needed to fulfill an incoming order
+* _payment-service_ executes the payment associated to an incoming order
+
+## Running the Example
+
+Build and start up:
+
+```console
+$ mvn clean verify
+```
+
+```console
+$ docker-compose up --build
+```
+
+Register the connectors for the different services:
+
+```console
+$ http PUT http://localhost:8083/connectors/order-outbox-connector/config < register-order-connector.json
+$ http PUT http://localhost:8083/connectors/payment-outbox-connector/config < register-payment-connector.json
+$ http PUT http://localhost:8083/connectors/credit-outbox-connector/config < register-credit-connector.json
+```
+
+Place an order:
+
+```console
+$ http POST http://localhost:8080/orders < requests/place-order.json
+```
+
+Examine the emitted messages for `payment` and `credit-approval` in Apache Kafka:
+
+```console
+$ docker run --tty --rm \
+    --network saga-network \
+    debezium/tooling:1.1 \
+    kafkacat -b kafka:9092 -C -o beginning -q \
+    -f "{\"key\":%k, \"headers\":\"%h\"}\n%s\n" \
+    -t payment.request
+
+$ docker run --tty --rm \
+    --network saga-network \
+    debezium/tooling:1.1 \
+    kafkacat -b kafka:9092 -C -o beginning -q \
+    -f "{\"key\":%k, \"headers\":\"%h\"}\n%s\n" \
+    -t credit-approval.request
+```
+
+Examine the saga state in the order service's database:
+
+```console
+$ docker run --tty --rm -i \
+        --network saga-network \
+        debezium/tooling:1.1 \
+        bash -c 'pgcli postgresql://orderuser:orderpw@order-db:5432/orderdb'
+
+select * from order.sagastate;
+
++--------------------------------------+------------------------------------------------------------------------------------------+----------+---------------------------------------------------+-----------------+-----------+
+| id                                   | payload                                                                                  | status   | stepstate                            | type            | version   |
+|--------------------------------------+------------------------------------------------------------------------------------------+----------+---------------------------------------------------+-----------------+-----------|
+| d08b4e43-8522-4539-9aba-de4bb0dc1a8e | {"payment-due":59,"customer-id":456,"order-id":2,"credit-card-no":"xxxx-yyyy-dddd-aaaa"} | COMPLETED  | {"credit-approval":"SUCCEEDED","payment":"SUCCEEDED"} | order-placement | 1         |
```

Comment:
I see an additional payload field `"type": "REQUEST"` in my demo run?

Reply:
also should the version column entry really be "1" in this case?

Reply:
for me it was version 3 for a successful request which is somewhat expected if this reflects to the saga states/transitions somehow?

Reply:
Good catch! Updated the example with the current payload. Regarding the version, you are right, it should have been 3 as per the state you reviewed. Now it's 2 actually (i.e. one statement less), as the code first did an INSERT + UPDATE when initially persisting the saga state, whereas now it's only the INSERT.